Add ssh key handling

itsbill opened the pull request:
Thank you for fast-vm!  I've come to use it nearly every day now.  There are two features I've wanted for some time and I've attempted to implement them here:  simplifying the creation of a separate ssh key, solely for fast-vm's use (for passwordless logins) and the ability to scp files easily.

* fast-vm keygen
Generate a passphraseless ssh key for fast-vm's use.  Stored in $FASTVM_USER_CONF_DIR

* fast-vm keydist <VmNumber|'VmNumber1 VmNumber2 ...'>
Distribute the fast-vm ssh key (using ssh-copy-id) to one or more VMs

* fast-vm scp <VmNumber|'VmNumber1 VmNumber2 ...'> /path/to/source/on/this/host [/path/on/destination/to/put/it]
Provide an easy way to scp files from the host to guest VMs, using the ssh key if it exists

* fast-vm ssh changes to utilize the fast-vm ssh key if it exists

OndrejHome commented:
Hi @itsbill,

Thank you for submitting the pull request! 
I will try to check this over weekend and let you know my thoughts.

OndrejHome commented on fast-vm:
```diff
@@ -212,6 +212,13 @@ calculate_day_difference () {
+use_sshkey_if_exists () {
+# If we our ssh key exists, set up to use it
+	if [ -f "$FASTVM_USER_CONF_DIR/id_fastvm" ]; then
+		FASTVM_USE_SSHKEY="-i $FASTVM_USER_CONF_DIR/id_fastvm -o BatchMode=yes"
```
Why we add the `-o BatchMode=yes`? 
When the key is found, but not present on VM, this will prevent us from logging into machine using just password.

itsbill replied:
Indeed you are correct.  It came from testing an earlier version and not needed.  Removed in the latest code, thanks.

OndrejHome commented on fast-vm:
```diff
@@ -966,9 +982,38 @@ case "$1" in
+		ssh-keygen -P '' -f $FASTVM_USER_CONF_DIR/id_fastvm
+
+		# work around FASTVM_USER_CONF_DIR being world accessible
+		chmod 0600 $FASTVM_USER_CONF_DIR/id_fastvm.pub
```
'public key' (`id_fastvm.pub`) being world readable is not as big problem as the 'private key' (`id_fastvm`) being readable. I wouldn't call it a workaround as the SSH requires the 'private key' to be accessible only by owner.

itsbill replied:
In testing I was getting an error from ssh, but I can no longer reproduce it.  Removed.

OndrejHome commented on man/fast-vm.8:
```diff
@@ -182,6 +195,15 @@ Number identifying the VM. Currently limited to range 20-220 to allow to be used
+.TP
+.I /path/on/destination/to/put/it
+Optional destination path on the VM to write the file being copied via SCP.
+If not specified, the default is the home directory.
```
The default is 'root user home directory' as we specify `root` for scp in the code. It will be better to make it clear here.

itsbill replied:
I've reworked this to not only do host-to-guest; you can now do:

Copy from host to guest:  fast-vm scp 52 afile vm:/tmp
Copy from guest to host:  fast-vm scp 52 vm:/tmp/afile .

OndrejHome commented on fast-vm:
```diff
@@ -966,9 +982,38 @@ case "$1" in
+			scp $FASTVM_USE_SSHKEY -o GlobalKnownHostsFile=/dev/null -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no "$sourcepath" "root@192.168.$SUBNET_NUMBER.$vm_number:$destpath"
+		fi
+		;;
+	keygen)
```
Is it (`fast-vm keygen`) really necessary to make this as another action?
To make this more convenient for user it would make more sense to me to generate the SSH key when user tries to do the `fast-vm ssh`, `fast-vm scp` or `fast-vm keydist` and there is no SSH key found.

itsbill replied:
You're right, in this version I have made it automatic, thanks.

OndrejHome commented on fast-vm:
```diff
@@ -966,9 +982,38 @@ case "$1" in
+			ssh $FASTVM_USE_SSHKEY -o GlobalKnownHostsFile=/dev/null -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no "root@192.168.$SUBNET_NUMBER.$vm_number"
 		fi
 		;;
+	scp|keydist)
```
I don't like having 2 actions together just to later use `if` to distinguish between them. This way we are loosing ability to do a proper argument check for SCP - where we can check if the `$3` (`sourcepath`) is a real file or not. I would really prefer to have these actions separately.

OndrejHome commented on fast-vm:
```diff
@@ -966,9 +982,38 @@ case "$1" in
+		if [ "$1" == "keydist" ]; then
+			ssh-copy-id -i "$FASTVM_USER_CONF_DIR/id_fastvm" \
+				-o GlobalKnownHostsFile=/dev/null -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no \
+				"root@192.168.$SUBNET_NUMBER.$vm_number" | sed "s/with:.*/with: $0 ssh $vm_number/"
```
What about filtering out more outputs and leave just 2 most important things - 'prompt for password' and 'information what command to use next'?
~~~
ssh-copy-id -i "$FASTVM_USER_CONF_DIR/id_fastvm" \
				-o GlobalKnownHostsFile=/dev/null -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no \
				"root@192.168.$SUBNET_NUMBER.$vm_number" 2>&1 |egrep '(password|Now try logging into the machine)'| sed "s/with:.*/with: $0 ssh $vm_number/"
~~~

itsbill replied:
Sure!  I changed the delimiter in sed to '@' to avoid problems with slashes in $0.

Please let me know what you think of the latest version.  I've updated the manual page as well.  Thank you!

OndrejHome reviewed (commented):
Please check my comments and questions to code. In general I like the idea here, but there are few places where I think we can change code a bit before this can be merged.

OndrejHome commented:
Thank you @itsbill for changes. Will try to review them over weekend and let you know if there are any additional changes needed or if it can be merged.

OndrejHome commented on fast-vm:
```diff
@@ -162,13 +162,18 @@ handle_multiple_requests () {
+			MULTI_OPERATION=1 $fast_vm_bin $operation $vm $original_arguments &
+		else
+			MULTI_OPERATION=1 $fast_vm_bin $operation $vm $original_arguments
+		fi
```
Why we serialize multiple operations (specifically the 'scp' and 'keydist')? Is this not working properly in parallel?

itsbill replied:
When a password is required, SERIALIZE allows fast-vm to iterate through the operation one VM at a time.  For keydist, a password will be required; for scp, if there is no ssh key, it would also be required.  Without SERIALIZE, the prompts appear to interfere with each other and the resulting output is confusing.  My thought was this would simplify distributing keys and files to multiple VMs at once.

OndrejHome commented on fast-vm:
```diff
@@ -907,10 +925,11 @@ case "$1" in
 						pmsg $P_ERROR "failed to start VM $vm_name, check syslog for more information\n" $vm_number
 					else
 						wait_for_ssh "$vm_number"
+						create_sshkey_if_needed
```
I don't like that SSH key will get generated just because we have done SSH to machine. Description of commit says `fast-vm ssh changes to utilize the fast-vm ssh key if it exists` which to me sounds like the SSH key is used only when found, not that it is automatically created. I would prefer the behaviour described in the commit message - in short it means that SSH key is generated only when needed - here it will be only in the `keydist` action.

itsbill replied:
Ah, I misinterpreted your earlier response, sorry.  In this version it is only created when when you call keydist and it doesn't already exist.

OndrejHome commented on fast-vm:
```diff
@@ -966,9 +985,38 @@ case "$1" in
 			fi
 		else
-			ssh -o GlobalKnownHostsFile=/dev/null -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no "root@192.168.$SUBNET_NUMBER.$vm_number"
+			create_sshkey_if_needed
```
same as comment on line 928

itsbill replied:
Removed.

OndrejHome commented on fast-vm:
```diff
@@ -966,9 +985,38 @@ case "$1" in
+
+		wait_for_ssh "$vm_number"
+		sourcepath=$(echo "$3" | sed "s/\(.*\)\bvm:\(.*\)/\1root@192.168.$SUBNET_NUMBER.$vm_number:\2/")
+		destpath=$(echo "$4" | sed "s/\(.*\)\bvm:\(.*\)/\1root@192.168.$SUBNET_NUMBER.$vm_number:\2/")
```
At least basic check that `$3` and `$4` are not empty would be good here, otherwise we receive the errors like
~~~
./fast-vm scp 100
cp: cannot stat '': No such file or directory

./fast-vm scp 100 /tmp/test
cp: cannot create regular file '': No such file or directory
~~~

itsbill replied:
Ah, I meant to add that check!  Added in this version, thanks:
$ ./fast-vm scp 51
[51][err] scp requires a sourcepath and destination path to copy
$ ./fast-vm scp 51 source
[51][err] scp requires a sourcepath and destination path to copy
$ ./fast-vm scp 51 source path
[51][inf] checking the 192.168.22.51 for active SSH connection (ctrl+c to interrupt)
[51][inf] 
[51]SSH ready
$

OndrejHome commented on fast-vm:
```diff
@@ -966,9 +985,38 @@ case "$1" in
+	keydist)
+		vm_number="$2"
+		if [ -z "$vm_number" ]; then pmsg $P_ERROR "VM number missing\n"; usage ssh; fi
+		if [[ $vm_number = *[\ ]* ]]; then handle_multiple_requests $@;	fi
```
I think that this function cannot be run in parallel so the `handle_multiple_request` should not be used here. I'm guessing that this is somehow connected with additional of SERIALIZE in the line 165. I would expect that if the operation requires output that cannot be automated (like here) then it wouldn't be possible to call it with multiple VMs.

itsbill replied:
You are correct -- SERIALIZE permits you to type the password for each VM that's receiving the key.  This is no worse than having to invoke fast-vm repeatedly, and maybe a little bit nicer?  :)

git will update shortly... thank you again!

OndrejHome reviewed (commented):
Hi @itsbill , Thank you for additional changes. Please check my notes on what I would expect to change/explain before this can be merged. Changes looks good and I have mostly minor comments, I was only surprised by the SERIALIZE option that I'm not sure why it is used. Could you please clarify that one?
(I would rather expect that `keydist` action cannot be run on multiple VMs because of interactive input it will for sure require)

Once we are sure on the parameters of all actions, there is also file `fast-vm.bash_completion` that add the bash_completion for actions. If you would like to change also that one please wait for confirmation from me on comments in this review. Also feel free to skip this and I will have a look at it once this is merged.

OndrejHome commented:
Hi @itsbill , its been some time since your last comment. Would you be looking into the above comments to address them in some near future (May/June)?

itsbill commented:
Hi @OndrejHome sorry for the delay.  I have made most of the required changes, but I want to doublecheck that I have addressed all the points you made earlier.  I hope to do this in the next week or so... thanks.

OndrejHome commented:
HI @itsbill, thank you for getting back. 
The SERIALIZE now starts to make sense to me. While it may cause not the most parallel execution I agree that it makes sense to make it not to mess the screen ( I was briefly considering making the SERIALIZE an option for some commands, but it doesn't seems at moment as best idea for the little gain that it would bring). 

I will pull down the changes over this/next weekend and test them out to see if everything was addressed.  If there are no huge changes needed I will merge this into master. 
I will see if I will create a release only with this change in master or if I will add some other changes and make bigger release in upcoming weeks. Stay tuned for updates :)

Thank you for taking the effort to provide this functionality!

OndrejHome commented:
Hi @itsbill , code is OK for me. I have some small changes below and one question. Otherwise works like a charm :+1: 

Question:
- Do you feel you would like to also update the `fast-vm.bash_completion` that contains bash completion? Current syntax is good for me so I think this is a logical step. If you prefer not to touch just let me know I can do that.

Small changes:
- warn about missing parameter for 'scp' _after_ checking validity of VM numbers - what happened to me was I tried command `fast-vm scp /local/file vm:remote_file` and I have received error about missing parameter but I should received the error on wrong VM number in my opinion
- change of man page formatting in some places so it looks pretty :)

~~~
diff --git a/fast-vm b/fast-vm
index 7873cc9..257f511 100755
--- a/fast-vm
+++ b/fast-vm
@@ -984,14 +984,14 @@ case "$1" in
                vm_number="$2"
                if [ -z "$vm_number" ]; then pmsg $P_ERROR "VM number missing\n"; usage scp; fi
                if [[ $vm_number = *[\ ]* ]]; then handle_multiple_requests $@; fi
+               check_vm_number "$vm_number" "inactive"
+               update_access_time "$vm_number"
+
                if [ -z "$4" ]; then
                        pmsg $P_ERROR "scp requires a sourcepath and destination path to copy\n" $vm_number
                        exit 1
                fi
 
-               check_vm_number "$vm_number" "inactive"
-               update_access_time "$vm_number"
-
                wait_for_ssh "$vm_number"
                sourcepath=$(echo "$3" | sed "s/\(.*\)\bvm:\(.*\)/\1root@192.168.$SUBNET_NUMBER.$vm_number:\2/")
                destpath=$(echo "$4" | sed "s/\(.*\)\bvm:\(.*\)/\1root@192.168.$SUBNET_NUMBER.$vm_number:\2/")
diff --git a/man/fast-vm.8 b/man/fast-vm.8
index 68f96b0..56f2894 100644
--- a/man/fast-vm.8
+++ b/man/fast-vm.8
@@ -117,8 +117,7 @@ fast-vm \(em script for defining VMs from images provided in thin LVM pool
 .B fast-vm
 .B scp
 .RI < VmNumber | "'VmNumber1 VmNumber2 ...'" >
-.RB vm:/remote/sourcefile|/local/sourcefile
-.RI /local/destfile|vm:/remote/destfile
+.RI < "vm:/remote/sourcefile /local/destfile" | "/local/sourcefile vm:/remote/destfile" >
 
 .SH DESCRIPTION
 fast-vm is provides command-line interface to create virtual machines (VMs) in 
@@ -364,7 +363,7 @@ Create machine with custom definition and hack file. Start it and after it's SSH
 
 .RB "scp a script to VMs 41 and 42. By default this goes in root's home directory, or specify a path instead"
 .sp
-.BI "fast-vm scp '41 42' script.sh vm:
+.BI "fast-vm scp " "'41 42' script.sh vm:"
 
 .SH EXIT CODES
 In case of error the fast-vm will return non-zero exit code. When multiple VMs were specified then zero exit code is returned only when operation succeeded on all VMs. If any of VMs reported non-zero exit code, then the overal exit code will also be non-zero.
~~~

itsbill commented:
Thank you @OndrejHome -- those changes look great!  I have not written bash completion before, but I try below.  I tried to have it add "vm:" to the source prompt for scp,and only add "vm:" to the destination completion if it wasn't already specified in the source.  This gets tricky because it contains a split char (colon)... anyway, here is my attempt, but please change as you see fit  :smile:

~~~
diff --git a/fast-vm.bash_completion b/fast-vm.bash_completion
index defd6f8..48a542d 100644
--- a/fast-vm.bash_completion
+++ b/fast-vm.bash_completion
@@ -94,6 +94,21 @@ _fast-vm()
 		return 0
 	fi
 
+	if [ "$cword" -eq 3 ] && [ ${COMP_WORDS[1]} == "scp" ]; then
+		COMPREPLY=( $( compgen -W "vm:" -f -- "$cur" ) )
+		return 0
+	fi
+	if [ "$cword" -eq 4 ] && [ ${COMP_WORDS[1]} == "scp" ]; then
+		# source path did /not/ contain "vm:", so prompt only that
+		COMPREPLY=( $( compgen -W "vm:" ) )
+		return 0
+	fi
+	if [ "$cword" -ge 5 ] && [ ${COMP_WORDS[1]} == "scp" ]; then
+		# source path /did/ contain "vm:", so exclude it from destination
+		COMPREPLY=( $( compgen -f -- "$cur" ) )
+		return 0
+	fi
+
 	case $prev in
 		import_image|import_custom_image|import_profile)
 			return 0
@@ -122,7 +137,7 @@ _fast-vm()
 			COMPREPLY=( $( compgen -W "all active inactive" -- "$cur" ) )
 			return 0
 			;;
-		start|stop|delete|console|ssh|info|resize)
+		start|stop|delete|console|keydist|scp|ssh|info|resize)
 			local vm_numbers
 			case $prev in
 				start|resize)
@@ -131,7 +146,7 @@ _fast-vm()
 				delete|edit_note|info)
 					vm_numbers=$(fast-vm list all short)
 				;;
-				console|ssh|stop)
+				console|keydist|scp|ssh|stop)
 					vm_numbers=$(fast-vm list active short)
 				;;
 			esac
@@ -142,7 +157,7 @@ _fast-vm()
 
 	$split && return 0
 	if [ "$cword" -lt 2 ]; then
-		COMPREPLY=( $( compgen -W 'import_image import_custom_image export_image remove_image resize_image import_profile remove_profile edit_note list_images list_profiles create start stop delete resize console ssh list info' -- "$cur" ) )
+		COMPREPLY=( $( compgen -W 'import_image import_custom_image export_image remove_image resize_image import_profile remove_profile edit_note keydist list_images list_profiles create start stop delete resize console scp ssh list info' -- "$cur" ) )
 		return 0
 	fi
~~~

Commit:
This commit enhances ssh-related operations:

- fast-vm keydist <VmNumber|'VmNumber1 VmNumber2 ...'>
  Distribute the fast-vm ssh key (using ssh-copy-id) to one or more VMs

keydist will automatically generate a passphraseless ssh key for
fast-vm's use, if it doesn't already exist, and store it in
$FASTVM_USER_CONF_DIR

If found, it will be used automatically for ssh and scp.  Can be run
against multiple VMs at once and will iterate through each in turn.

- fast-vm scp <VmNumber|'VmNumber1 VmNumber2 ...'> <vm:/remote/sourcefile /local/sourcefile|/local/destfile vm:/remote/destfile>

Provide an easy way to scp files between the host and guest VMs,
using the ssh key if it exists.  "vm" as a hostname in the source
or destination will be replaced with the VM addresses specified,
allowing easy copy-to-vm or copy-from-vm interaction.

- fast-vm ssh changes to utilize the fast-vm ssh key if it exists

itsbill commented:
I reset and incorporated your diff into this latest version, thanks.  (I left out the bash completion diff because I am not confident that it is good enough.)
Thanks again!

OndrejHome commented:
Thanks a lot @itsbill ! I will check this over this weekend.

OndrejHome commented:
I will play with bash completion tomorrow. The changes are now merged in master, but I will try to finish other tasks to release it together with version 1.4.0 soon.
